File: converter/regress_render_contract.mjs
// Render-contract measurements: real Yomitan generator (jsdom) -> real Chrome.
// Driven by converter/regress_render_contract.py, which supplies the payload and
// the stylesheet and then asserts on the JSON printed here.
import fs from 'node:fs';
import path from 'node:path';
import {fileURLToPath, pathToFileURL} from 'node:url';
import {spawnSync} from 'node:child_process';
import {JSDOM, VirtualConsole} from '../scgen_test/node_modules/jsdom/lib/api.js';
import {StructuredContentGenerator} from '../scgen_test/js/display/structured-content-generator.js';

const here = path.dirname(fileURLToPath(import.meta.url));
const scratch = path.join(here, '_rc');
fs.mkdirSync(scratch, {recursive: true});
const payloadPath = process.argv[2];
const cssPath = process.argv[3];
const payload = JSON.parse(fs.readFileSync(payloadPath, 'utf8'));
const css = fs.readFileSync(cssPath, 'utf8');

const dom = new JSDOM('<!doctype html><html><head><meta charset="utf-8"></head><body></body></html>',
                      {virtualConsole: new VirtualConsole()});
const win = dom.window;
globalThis.location = win.location;
class Fake {
    prepareLink(n, h) { n.setAttribute('href', h); }
    prepareScripts() {} prepareHTML() {} loadMedia() {} openMediaInTab() {}
}
const gen = new StructuredContentGenerator(new Fake(), win.document, win);

// light + dark exercise the theme palette (the R3 regression); no-css exercises
// what an Anki export / plain-HTML host actually sees.
const modes = ['light', 'dark', 'no-css'];
for (const mode of modes) {
    const section = win.document.createElement('section');
    section.dataset.mode = mode;
    const light = (mode === 'light' || mode === 'no-css');
    if (mode !== 'no-css') section.className = 'styled';
    section.style.cssText =
        `--text-color:${light ? '#000000' : '#d4d4d4'};` +
        `background:${light ? '#ffffff' : '#1e1e1e'};color:var(--text-color);padding:16px`;
    for (const [word, sc] of Object.entries(payload)) {
        const holder = win.document.createElement('div');
        holder.dataset.word = word;
        holder.append(gen.createStructuredContent(sc, 'LDOCE5pp (LM5pp)'));
        section.append(holder);
    }
    win.document.body.append(section);
}
const style = win.document.createElement('style');
style.textContent = `body{margin:0;font:16px/1.5 sans-serif}.styled{${css}}`;
win.document.head.append(style);

const script = win.document.createElement('script');
script.textContent = String.raw`
const rgb = c => { const cv=document.createElement('canvas'); cv.width=cv.height=1;
  const ctx=cv.getContext('2d'); ctx.fillStyle=c; ctx.fillRect(0,0,1,1);
  return [...ctx.getImageData(0,0,1,1).data].slice(0,3); };
const lum = a => a.map(x => { x/=255; return x<=.04045 ? x/12.92 : ((x+.055)/1.055)**2.4; })
  .reduce((s,x,i) => s + x*[.2126,.7152,.0722][i], 0);
const contrast = (a,b) => { a=lum(a); b=lum(b);
  return +((Math.max(a,b)+.05)/(Math.min(a,b)+.05)).toFixed(2); };
// first visible text leaf inside a node: inline styles land on descendants, so
// reading the wrapper's own colour would miss the declaration we care about
const leafOf = n => { const w=document.createTreeWalker(n,NodeFilter.SHOW_TEXT); let t;
  while ((t=w.nextNode()) && !t.textContent.trim()) {}
  return t ? t.parentElement : n; };
const out = { userAgent: navigator.userAgent,
              colorMix: CSS.supports('color','color-mix(in srgb,red,blue)'), modes: {} };
const CLASSES = ['ld-pos','ld-gram','ld-defcn','ld-excn','ld-field','ld-grouptitle'];
for (const section of document.querySelectorAll('section[data-mode]')) {
  const bg = rgb(getComputedStyle(section).backgroundColor);
  const rec = {fields:{}, weights:{}, marks:{}, lists:[], defMarginBottom:null};
  const host = section.querySelector('[data-word="act"]') || section;
  for (const cls of CLASSES) {
    const n = section.querySelector('[data-sc-class="'+cls+'"]');
    if (!n) continue;
    const leaf = leafOf(n), cs = getComputedStyle(leaf);
    rec.fields[cls] = { inline: n.style.color || null, computed: cs.color,
                        contrast: contrast(rgb(cs.color), bg),
                        text: (leaf.textContent||'').trim().slice(0,40) };
  }
  for (const cls of ['ld-nodew','ld-colloin','ld-collo']) {
    const n = section.querySelector('[data-sc-class="'+cls+'"]');
    if (n) rec.weights[cls] = { inline: n.style.fontWeight || null,
                                computed: getComputedStyle(n).fontWeight };
  }
  for (const cls of ['ld-mark','ld-snum','ld-empty']) {
    const n = section.querySelector('[data-sc-class="'+cls+'"]');
    rec.marks[cls] = n ? { display: getComputedStyle(n).display,
                           text: (n.textContent||'').trim().slice(0,12),
                           fontSize: getComputedStyle(n).fontSize,
                           width: +n.getBoundingClientRect().width.toFixed(2) } : null;
  }
  for (const ol of section.querySelectorAll('[data-sc-class="ld-senselist"]')) {
    const snums = [...ol.querySelectorAll('[data-sc-class="ld-snum"]')]
      .filter(n => n.closest('[data-sc-class="ld-senselist"]') === ol);
    const lis = [...ol.children].filter(c => c.tagName === 'LI');
    rec.lists.push({ listStyle: getComputedStyle(ol).listStyleType,
                     start: ol.start, items: lis.length,
                     sourceNumbers: snums.map(n => (n.textContent||'').trim()),
                     chipFontSizes: snums.map(n => getComputedStyle(n).fontSize),
                     chipWidths: snums.map(n => +n.getBoundingClientRect().width.toFixed(2)),
                     inlineListStyle: ol.style.listStyleType || null,
                     nested: [...ol.querySelectorAll('ul')].length });
  }
  const d = host.querySelector('[data-sc-class="ld-def"]');
  if (d) rec.defMarginBottom = getComputedStyle(d).marginBottom;
  out.modes[section.dataset.mode] = rec;
}
const pre = document.createElement('pre');
pre.id = 'contract-result';
pre.textContent = JSON.stringify(out);
document.body.prepend(pre);
`;
win.document.body.append(script);

const html = path.join(scratch, 'page.html');
fs.writeFileSync(html, dom.serialize(), 'utf8');
const chrome = 'C:/Program Files/Google/Chrome/Application/chrome.exe';
const r = spawnSync(chrome, ['--headless=new', '--no-sandbox', '--disable-gpu',
    '--no-first-run', '--no-default-browser-check', '--disable-extensions',
    '--disable-background-networking', '--disable-sync',
    '--user-data-dir=' + path.join(scratch, 'chrome-profile'),
    '--dump-dom', pathToFileURL(html).href],
    {encoding: 'utf8', timeout: 120000, maxBuffer: 40 * 1024 * 1024, windowsHide: true});
if (r.status !== 0) {
    console.error('chrome failed', r.status, (r.stderr || '').slice(0, 500));
    process.exit(r.status || 2);
}
const rendered = new JSDOM(r.stdout, {virtualConsole: new VirtualConsole()});
const node = rendered.window.document.querySelector('#contract-result');
if (!node) {
    console.error('chrome produced no measurement element');
    process.exit(3);
}
process.stdout.write(node.textContent);

File: converter/regress_render_contract.py
"""Render-contract gate: real Yomitan generator + headless Chrome.

Asserts the two environments behave as designed, on the ACTUAL package content:

  with styles.css   -- the theme palette wins over the inline fallbacks. Concretely
                       ld-pos / ld-gram must not render DodgerBlue and ld-defcn /
                       ld-excn / ld-field / ld-grouptitle must not render green;
                       each must clear 4.3:1 against its own card background in both
                       the light and the dark theme. ld-nodew / ld-colloin must be
                       600, not the 700 the fallback carries. (audit R3)
  without styles.css -- the UA list marker stays OFF so the number the reader sees is
                       the SOURCE number: 'act' must show 7,8,9,10 (never 1,2,3,4),
                       the chip must be visible, and the inline marker spans that
                       carry the example dashes / ticks must be visible. (audit R1/R2)

Usage:
    python converter/regress_render_contract.py [term_bank_dir_or_zip]
"""
import json
import os
import re
import subprocess
import sys
import zipfile

HERE = os.path.dirname(os.path.abspath(__file__))
ROOT = os.path.dirname(HERE)
sys.path.insert(0, HERE)
import ldoce2yomitan as C  # noqa: E402

WORDS = ["act", "access", "above", "criminal", "7/7", "bad", "the", "matter"]
NEEDED = ["ld-pos", "ld-defcn", "ld-excn"]
MIN_CONTRAST = 4.3


def load_rows(src):
    if src.endswith(".zip"):
        with zipfile.ZipFile(src) as z:
            names = sorted((n for n in z.namelist() if re.fullmatch(r"term_bank_\d+\.json", n)),
                           key=lambda n: int(re.search(r"\d+", n).group()))
            for n in names:
                yield from json.loads(z.read(n))
    else:
        for name in sorted(os.listdir(src)):
            if re.fullmatch(r"term_bank_\d+\.json", name):
                with open(os.path.join(src, name), encoding="utf-8") as fh:
                    yield from json.load(fh)


def main():
    src = sys.argv[1] if len(sys.argv) > 1 else os.path.join(ROOT, "_smoke3")
    payload = {}
    for row in load_rows(src):
        if row[4] > 0 and row[0] in WORDS and row[0] not in payload:
            payload[row[0]] = row[5][0]["content"]
    missing = [w for w in WORDS if w not in payload]
    if missing:
        print(f"note: source has no content row for {missing}")

    payload_path = os.path.join(HERE, "_rc", "payload.json")
    css_path = os.path.join(HERE, "_rc", "styles.css")
    os.makedirs(os.path.dirname(payload_path), exist_ok=True)
    with open(payload_path, "w", encoding="utf-8") as fh:
        json.dump(payload, fh, ensure_ascii=False)
    # Audit the stylesheet actually shipped with the supplied content. Using
    # generate_css() here can make an old, broken ZIP pass after a source-only fix.
    if src.endswith(".zip"):
        with zipfile.ZipFile(src) as z:
            css = z.read("styles.css").decode("utf-8")
    else:
        with open(os.path.join(src, "styles.css"), encoding="utf-8") as fh:
            css = fh.read()
    with open(css_path, "w", encoding="utf-8", newline="\n") as fh:
        fh.write(css)

    proc = subprocess.run([C.shutil.which("node") or "node",
                           os.path.join(HERE, "regress_render_contract.mjs"),
                           payload_path, css_path],
                          capture_output=True, text=True, encoding="utf-8", cwd=ROOT)
    if proc.returncode != 0:
        print("node failed:", proc.returncode)
        print(proc.stderr[:2000])
        return 2
    data = json.loads(proc.stdout)

    fails = []

    def check(ok, label, detail=""):
        print(f"  {'OK ' if ok else 'FAIL'} {label}" + (f"  {detail}" if detail else ""))
        if not ok:
            fails.append(label + (" " + detail if detail else ""))

    print(f"chrome: color-mix supported = {data['colorMix']}")
    print("\n== with styles.css ==")
    for mode in ("light", "dark"):
        rec = data["modes"][mode]
        print(f"-- {mode} --")
        for cls in NEEDED:
            if cls not in rec["fields"]:
                fails.append(f"{mode}: {cls} not present in the sample")
                continue
            f = rec["fields"][cls]
            bad = f["computed"] in ("rgb(30, 144, 255)", "rgb(0, 128, 0)")
            check(not bad and f["contrast"] >= MIN_CONTRAST,
                  f"{mode} {cls:11}", f"colour={f['computed']:20} inline={f['inline']!s:12} "
                                     f"contrast={f['contrast']}")
        for cls, want in (("ld-nodew", "600"), ("ld-colloin", "600")):
            if cls in rec["weights"]:
                w = rec["weights"][cls]
                check(w["computed"] == want, f"{mode} {cls} weight",
                      f"computed={w['computed']} inline={w['inline']}")
        for cls, want in (("ld-mark", "none"), ("ld-snum", None)):
            m = rec["marks"].get(cls)
            if not m:
                continue
            if want:
                check(m["display"] == want, f"{mode} {cls} display", f"={m['display']}")
            else:
                check(m["display"] != "none" and m["fontSize"] != "0px" and m["width"] > 0,
                      f"{mode} {cls} visible",
                      f"display={m['display']} fontSize={m['fontSize']} w={m['width']}")
        for ol in rec["lists"]:
            check(ol["listStyle"] == "none", f"{mode} ol list-style", f"={ol['listStyle']}")
            check("0px" not in ol["chipFontSizes"], f"{mode} ol chip not collapsed",
                  f"fontSizes={ol['chipFontSizes'][:4]}")
        check(rec["defMarginBottom"] == "1px", f"{mode} ld-def margin-bottom",
              f"={rec['defMarginBottom']}")

    print("\n== with CSS, without Yomitan theme variables ==")
    for mode in ("no-var-light", "no-var-dark"):
        body = data["modes"][mode].get("plainText")
        check(body is not None, f"{mode} plain definition present")
        if body is not None:
            check(body["themeVariable"] == "", f"{mode} has no --text-color")
            check(body["rootColor"] == body["hostColor"]
                  and body["definitionColor"] == body["hostColor"],
                  f"{mode} root and definition inherit the host",
                  f"root={body['rootColor']} host={body['hostColor']}")
            check(body["contrast"] >= MIN_CONTRAST,
                  f"{mode} plain definition remains readable", f"contrast={body['contrast']}")

    print("\n== without styles.css ==")
    rec = data["modes"]["no-css"]
    for ol in rec["lists"]:
        check(ol["listStyle"] == "none",
              "no-css ol marker suppressed", f"inline={ol['inlineListStyle']} computed={ol['listStyle']}")
        if ol["sourceNumbers"]:
            # 49% of senses carry no number at all, so only lists that DO have a
            # chip are asked to render it
            check(ol["chipWidths"] and min(ol["chipWidths"]) > 0,
                  "no-css chip visible", f"widths={ol['chipWidths'][:6]}")
        src_nums = ol["sourceNumbers"]
        if src_nums[:1] == ["7"]:
            check(src_nums[:4] == ["7", "8", "9", "10"],
                  "no-css act shows the SOURCE numbers", f"{src_nums[:6]}")
    mark = rec["marks"].get("ld-mark")
    if mark:
        check(mark["display"] != "none", "no-css marker span visible", f"display={mark['display']}")
    # with no stylesheet the literal fallback is what MUST show -- that is the whole
    # point of SEMANTIC_INLINE_STYLES
    FALLBACK = {"ld-pos": "rgb(30, 144, 255)", "ld-gram": "rgb(30, 144, 255)",
                "ld-defcn": "rgb(0, 128, 0)", "ld-excn": "rgb(0, 128, 0)",
                "ld-field": "rgb(0, 128, 0)", "ld-grouptitle": "rgb(0, 128, 0)"}
    for cls, want in FALLBACK.items():
        f = rec["fields"].get(cls)
        if f:
            check(f["computed"] == want, f"no-css {cls} shows the fallback colour",
                  f"={f['computed']} (want {want})")

    print()
    if fails:
        print(f"RESULT: FAIL ({len(fails)})")
        for f in fails:
            print("  -", f)
        return 1
    print("RESULT: PASS -- palette/host inheritance with CSS; source numbering and markers without")
    return 0


if __name__ == "__main__":
    sys.exit(main())
